src/v/cluster/topics_frontend.cc

```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
     if (!units) {
         co_return make_error_result(assignable_config.cfg.tp_ns, units.error());
     }
+
+    // Enforce cluster-wide topic count limit only on user created topics.
+    auto max_user_topics_opt = _max_user_topics();
+    if (max_user_topics_opt.has_value()) {
+        // This is intended to approximate the number of topics that are
+        // user-defined and may return a count slightly smaller than the actual
+        // amount.
+        auto user_topic_count = std::max(
+                                  _topics.local().all_topics_count(),
+                                  model::non_user_topics.size())
+                                - model::non_user_topics.size();
+        if (
+          user_topic_count > *max_user_topics_opt
+          && model::is_user_topic(tp_ns)) {
+            co_return make_error_result(
+              assignable_config.cfg.tp_ns,
+              make_error_code(errc::topic_operation_error),
```

Comment:
Does this give a good error at the Kafka side, same as the error you get if you hit a partition limit, including a good detailed error message?

Reply:
```
topic_invalid_partitions_core_limit,
topic_invalid_partitions_memory_limit,
topic_invalid_partitions_fd_limit,
topic_invalid_partitions_decreased
```

Those look suspiciously like Redpanda error reasons, I think the Kafka error is just a general one like `INVALID_PARTITIONS`. You can use the same error code here, it's just important that a good error _string_ is sent back and is reported in a good way by rpk and other tools. When the partition limits went in we went though this and had to eventually add that because there was too much confusion around why topic creation requests failed.